### ISSUE TITLE
Support more platforms

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/canonical-data-syncer#24]
In particular, we should be supporting:

- Linux 32-bit
- Windows 32-bit
- ARM

**Comment by ee7** [comment]
All we can say from the current configlet download data is that it's downloaded about 3x more often on macOS than Windows. We can't easily know the Linux numbers, since the vast majority are CI downloads.

But it's interesting to look at the stats for [the most recent release of `exercism/cli`](https://github.com/exercism/cli/releases/tag/v3.0.13), which gives us download data from the last 2.5 years. 

```shell
gh release view v3.0.13 --repo exercism/cli --json assets \
  | jq -r '.assets | .[] | [.name,.downloadCount] | @tsv' \
  | column -t -N 'Asset name,Download count' -R 2 -o ' | '
```

| Asset name                            | Download count |
| ------------------------------------- | --------------:|
| exercism-3.0.13-darwin-i386.tar.gz    |           1613 |
| exercism-3.0.13-darwin-x86_64.tar.gz  |           3187 |
| exercism-3.0.13-freebsd-i386.tar.gz   |             56 |
| exercism-3.0.13-freebsd-x86_64.tar.gz |            275 |
| exercism-3.0.13-linux-armv5.tar.gz    |            211 |
| exercism-3.0.13-linux-armv6.tar.gz    |            726 |
| exercism-3.0.13-linux-i386.tar.gz     |            342 |
| exercism-3.0.13-linux-ppc64.tar.gz    |            153 |
| exercism-3.0.13-linux-x86_64.tar.gz   |          24161 |
| exercism-3.0.13-openbsd-i386.tar.gz   |             52 |
| exercism-3.0.13-openbsd-x86_64.tar.gz |            212 |
| exercism-3.0.13-windows-i386.zip      |            683 |
| exercism-3.0.13-windows-x86_64.zip    |           7053 |
| exercism-freebsd-32bit.tgz            |             44 |
| exercism-freebsd-64bit.tgz            |            155 |
| exercism-linux-32bit.tgz              |            225 |
| exercism-linux-64bit.tgz              |          21820 |
| exercism-linux-ppc64.tgz              |             70 |
| exercism-mac-32bit.tgz                |            191 |
| exercism-mac-64bit.tgz                |           3938 |
| exercism-openbsd-32bit.tgz            |             33 |
| exercism-openbsd-64bit.tgz            |            218 |
| exercism-windows-32bit.zip            |           1907 |
| exercism-windows-64bit.zip            |         115582 |
| exercism_checksums.txt                |            631 |
| exercism_checksums.txt.sig            |            126 |

After combining duplicate release data (the release naming convention was changed, but releases with the old convention were also uploaded for backward compatibility), then sorting:

| Operating System | Architecture | Download count |
| ---------------- | ------------ | --------------:|
| Windows          | x86_64       |         122635 |
| Linux            | x86_64       |          45981 |
| macOS            | x86_64       |           7125 |
| Windows          | i386         |           2590 |
| macOS            | i386         |           1804 |
| Linux            | armv6        |            726 |
| Linux            | i386         |            567 |
| FreeBSD          | x86_64       |            430 |
| OpenBSD          | x86_64       |            430 |
| Linux            | ppc64        |            223 |
| Linux            | armv5        |            211 |
| FreeBSD          | i386         |            100 |
| OpenBSD          | i386         |             85 |

I imagine that:
- The OS/Arch data for "people who use configlet" is vastly different from "people who use the Exercism CLI".
- The Exercism CLI usage changed drastically after the release of Exercism v3, and some types of user are more likely to keep using the CLI.

But still... interesting.

I'm hoping that we can release configlet on more platforms than the CLI (adding arm64 on Linux and macOS at least), but if there's some unforeseen reason that we have to prioritise certain platforms, it might be worth keeping the above data in mind.

Note that I have no plans to set up CI pipelines for platforms that GitHub Actions do not provide - the others will be cross-compiled and largely untested (although we could use https://github.com/docker/setup-qemu-action). Azure got arm64 recently, so GitHub Actions might get that eventually too. Otherwise, BSDs are available on sourcehut, and ARM/ARM64 are available on Drone CI and CircleCI. (And in the days before GitHub Actions, I used to use arm64, ppc64, and s390x on Travis CI).

Let me know if you have any thoughts.

**Comment by ErikSchierboom** [comment]
> adding arm64 on Linux and macOS at least

Agreed. Those seem to be the ones that would benefit most from having a specific build.

> Note that I have no plans to set up CI pipelines for platforms that GitHub Actions do not provide

I agree with that. We can put in a caveat in the README or something that it is untested, but with the current complexity of the code, I don't think it is likely that there'll be many issues. I'd prefer _not_ to have different CI platforms.

**Comment by kotp** [comment]
Anyway to measure builds?  Those that compile the CLI instead of using a pre-compiled executable?

**Comment by ee7** [comment]
> Anyway to measure builds? Those that compile the CLI instead of using a pre-compiled executable?

If you mean "can we know how many times has `go build` been run to compile `exercism/cli`?", or "how many people are using the CLI without using a CLI release binary?" then I think the answer is "no". I imagine that the build process for the CLI release binaries doesn't do something to make those binaries distinguishable later from self-built ones, but I could be wrong about that.

But there is both a binary and source package in e.g. the AUR:
- https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=exercism-bin
- https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=exercism

There are some `git clone` stats at https://github.com/exercism/cli/graphs/traffic for those with certain privileges in that repo, but the stats are only for the previous 2 weeks, and they might be too noisy to be useful.

And AFAIK, GitHub does not make download stats available for the source archives (e.g. https://github.com/exercism/cli/archive/refs/tags/v3.0.13.tar.gz).

**Comment by kotp** [comment]
There could be a "call home" trigger in the build that could report this.  Not sure if it is worthwhile doing though.

**Comment by ee7** [comment]
I think the plan should be something like:

First, get `x86-64` and `arm64` on Linux/macOS/Windows:

```text
configlet_4.1.0_linux_arm64.tar.gz
configlet_4.1.0_linux_x86-64.tar.gz
configlet_4.1.0_macos_arm64.tar.gz
configlet_4.1.0_macos_x86-64.tar.gz
configlet_4.1.0_windows_arm64.zip
configlet_4.1.0_windows_x86-64.zip
```

Then add BSDs, and more 64-bit Linux builds: 

```text
configlet_4.2.0_freebsd_x86-64.tar.gz
configlet_4.2.0_linux_powerpc64.tar.gz
configlet_4.2.0_linux_riscv64.tar.gz
configlet_4.2.0_netbsd_x86-64.tar.gz
configlet_4.2.0_openbsd_x86-64.tar.gz
```

Then consider adding 32-bit assets, like:

```text
configlet_4.3.0_linux_arm.tar.gz
configlet_4.3.0_linux_i386.tar.gz
configlet_4.3.0_windows_i386.zip
```

(Where the version numbers have no particular meaning).

**Comment by ee7** [comment]
@ErikSchierboom you wrote (albeit nearly 3 years ago):
 
>In particular, we should be supporting:
>
>- Linux 32-bit
>- Windows 32-bit

Did this mean:

1. x86
1. arm
1. x86 and arm
1. or x86, arm, plus others?

And how much do we care about this? Was it just motivated by [the assets we had for v2-configlet](https://github.com/exercism/v2-configlet/releases/tag/v3.9.2)?

It looks like `zig cc` will [eventually have Tier 1 support for](https://ziglang.org/download/0.11.0/release-notes.html#Support-Table):

- `arm-windows-gnu`
- `arm-linux-musleabi`
- `mipsel-linux-musl`
- `powerpc-linux-musl`
- `x86-linux-musl`
- `x86-windows-gnu`

But I'm not keen on providing builds for them because:

- I suspect that the number of people who currently (or in the future) want to run configlet on 32-bit platforms is essentially 0.
- We'd need some dedicated changes, because e.g. some prob-specs JSON files (e.g. for `grains`) contain numbers that exceed the maximum value of a 32-bit int.
- It's simplest for configlet to have the absolute minimum amount of architecture-specific code.
- It's questionable to provide those executables pre-built if we don't have integration tests for them.
- Adding 32-bit integration testing will take a little bit of work (but probably not much if we gain qemu integration testing of the existing cross-compiled executables).
- It'd be better to spend time on any other configlet issue.

Thoughts?

Adding further platforms isn't a priority for me. But if I do it for my own interest, next would be x86_64 BSDs (I've built and tested configlet on them in the past) and more 64-bit Linux builds (e.g. powerpc64).

**Comment by ErikSchierboom** [comment]
> And how much do we care about this? Was it just motivated by [the assets we had for v2-configlet](https://github.com/exercism/v2-configlet/releases/tag/v3.9.2)?

It was, so I'm happy to ditch that. We've also hadn't really had people requesting 32-bit support AFAICT so 🤷

**Comment by ee7** [comment]
Okay, then I'll suggest that this issue is closed by our recent adding of arm64 assets in these commits:

- https://github.com/exercism/configlet/commit/0e8d6659e43dbf0628abcd5ba477e0a011a7058a
- https://github.com/exercism/configlet/commit/f280445978c16e9590d3822f5795b1485a64a65f
- https://github.com/exercism/configlet/commit/3824299df1dda6b020a22c21a18808a5da29f757

I'll open separate issues to track some possible future additions like the BSDs, but won't open issues for 32-bit targets.
